File: Task-03/Sub-Task 4/Ruby/Ques_4.rb
def generate_diamond(n)
  diamond = []

  (1..n).step(2) do |i|
    spaces = ' ' * ((n - i) / 2)
    stars = '*' * i
    diamond << spaces + stars
  end

  (n - 2).step(1, -2) do |i|
    spaces = ' ' * ((n - i) / 2)
    stars = '*' * i
    diamond << spaces + stars
  end

  diamond
end

n = File.read('input.txt').to_i

diamond_pattern = generate_diamond(n)

File.open('output.txt', 'w') do |file|
  file.puts diamond_pattern
end

puts "Diamond pattern written to output.txt"

File: Task-03/Sub-Task 4/Rust/Ques_4.rs
use std::fs;

fn generate_diamond(n: usize) -> Vec<String> {
    let mut diamond = Vec::new();

    for i in (1..=n).step_by(2) {
        let spaces = " ".repeat((n - i) / 2);
        let stars = "*".repeat(i);
        diamond.push(format!("{}{}", spaces, stars));
    }

    for i in (1..n).rev().step_by(2) {
        let spaces = " ".repeat((n - i) / 2);
        let stars = "*".repeat(i);
        diamond.push(format!("{}{}", spaces, stars));
    }

    diamond
}

fn main() {
    let n: usize = fs::read_to_string("input.txt")
        .expect("Unable to read input.txt")
        .trim()
        .parse()
        .expect("Input is not a valid number");

    let diamond_pattern = generate_diamond(n);

    fs::write("output.txt", diamond_pattern.join("\n"))
        .expect("Unable to write to output.txt");

    println!("Diamond pattern written to output.txt");
}
